### PR TITLE
refactor: liquidation logic

### PR DESCRIPTION
**Pull request by yrong**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

**Comment by yrong** [comment]
mainly follow items in improve list  
    ```
1. liquidationAsset vs debtAsset, liquidationAssetReserve vs debtReserve
        a. if we really want to allow non-borrowed asset for liquidation, we should always use `liquidationAsset` instead of debtAsset OR debtReserve
2. baseCollateral naming is very confusing, we can remove it and keep only `maxCollateralToLiquidate`
3. collateralAmount naming is very confusing, can it be renamed as `actualCollateralToLiquidate`
4. collateralDiscountedPrice is very confusing, can it be renamed as
    
    `actualLiquidationAmount`
    
5. erc721 liquidation shouldn’t read decimals ⇒ to keep consistency because other places we didn’t use
6. doesn’t respect CEI(check effects interaction)
```

for item 1 not change yet , for me both are ok  but prefer `debtAsset` from Aave a little bit more easy to understand than `liquidationAsset` , but I'm ok if we all agree to change @WalidOfNow @GopherJ

**Comment by GopherJ** [comment]
@yrong the issue is that we support liquidation using non-borrowed assets

File: contracts/protocol/libraries/logic/LiquidationLogic.sol [hunk 1]
```diff
@@ -344,6 +339,12 @@ library LiquidationLogic {
             vars.actualDebtToLiquidate
         );
 
+        if (params.receiveXToken) {
```
Comment (zhoujia6139) [inline contracts/protocol/libraries/logic/LiquidationLogic.sol:342]: can we also move burn debtToken logic here.

```
        if (vars.userTotalDebt == vars.actualDebtToLiquidate) {
            userConfig.setBorrowing(debtReserve.id, false);
        }
        _burnDebtTokens(params, vars);
```

**Comment by yrong** [comment]
> @yrong the issue is that we support liquidation using non-borrowed assets

@GopherJ so you mean in some cases debt asset to be in liquidation may not borrowed directly by the user, so better to use `liquidationAsset` to replace, right?

**Review by WalidOfNow — APPROVED** [review]

**Review by GopherJ — APPROVED** [review]

**Pull request merged**